### PR TITLE
add gRPC Transport for Vmess/trojan

### PR DESCRIPTION
**Pull request by ShinyGwyn**
https://github.com/Qv2ray/gun Thank @DuckSoft @studentmain 
v2fly/v2ray-core#757 Thank  @xiaokangwang

### CODE REVIEW

**Comment by Dreamacro** [comment]
gRPC (protobuf) too heavy-weight.

Take linux-amd64 as an example. On dev branch, the binary size is 8.8M, but on this PR, the binary size is 13M. Increased size by nearly 50% 🌚

**Comment by ShinyGwyn** [comment]
Okay. I just want it to be merged into Premium. Maybe I don't have to close this PR

**Comment by Loyalsoldier** [comment]
Why to care about binary size 😂

**Comment by Dreamacro** [comment]
@Loyalsoldier From the beginning, the clash has been trying to control the size on purpose. As you can see in `go.mod`, the "heaviest" library in there is probably yaml.

As for http framework, I chose the lightest library `github.com/go-chi/chi`. Neither gin nor fiber.

**Comment by DuckSoft** [comment]
This protobuf is very simple, so I think we can emulate the same effect by reusing some HTTP/2 library and pretend to be a gRPC-compliant client... I am working on it and will post here if any update is available.

**Comment by Dreamacro** [comment]
@ShinyGwyn @DuckSoft It would be great if this transport could be implemented without importing a large library

**Comment by DuckSoft** [comment]
Done. Very crude code: https://github.com/Qv2ray/gun-lite, but usable.
Thanks github.com/ginuerzh/gost (MIT).

**Comment by Dreamacro** [comment]
@ShinyGwyn
It looks like you can't use https://github.com/Qv2ray/gun-lite directly.
On [Qv2ray/gun-lite/pkg/realgun/conn.go#L43-L65](https://github.com/Qv2ray/gun-lite/blob/b1ea1629929ffb5ad772af8f9fb1cc02bd935238/pkg/realgun/conn.go#L43-L65), It defines a specific `DialFn` function that the clash can't dial tcp with some special behavior (e.g. bind a interface).
This is why every component (e.g. [github.com/Dreamacro/clash/component/trojan/trojan.go#L49](https://github.com/Dreamacro/clash/blob/6c83ff3496542f12297eb971a611d3d3aea90297/component/trojan/trojan.go#L49)) in the clash is designed with a function like `StreamConn`.

Maybe we should move the code directly into the clash for the necessary modifications. What do you think? @DuckSoft

**Comment by DuckSoft** [comment]
@Dreamacro I am totally fine if you merge the code into Clash codebase. Just leave a comment like this at the beginning of that file is enough:
```
// Modified from: https://github.com/Qv2ray/gun-lite
// License: MIT
```

**Comment by Dreamacro** [comment]
@ShinyGwyn @DuckSoft Can you see if there are any mistakes in the current changes? The current `ProxyAdapter` interface doesn't support mux connections very clearly and will be improved after this pr merge.

**Review by DuckSoft — APPROVED** [review]
Looks good to me. Let @ShinyGwyn have a test and we are ready to go.

**Comment by ShinyGwyn** [comment]
I'm all right with that, since I didn't contribute most of the code. But I tested it and the current code is not working

**Comment by Dreamacro** [comment]
@ShinyGwyn Can you provide a copy of the configuration you tested?

**Comment by ShinyGwyn** [comment]
> @ShinyGwyn Can you provide a copy of the configuration you tested?

```
- {name: test-grpc, server: grpc.examples.com, port: 443, type: trojan, password: "examples", network: grpc, udp: true, grpc-service-name: "examples" }
```

**Comment by Dreamacro** [comment]
@ShinyGwyn What I want is trojan server configuration 😂

**Comment by ShinyGwyn** [comment]
```
{
      "port": 443,
      "listen": "0.0.0.0",
      "protocol": "trojan",
      "settings": {
          "clients": [
              {
                  "password": "example",
                  "email": "grpc@example.com"
              }
          ],
          "fallbacks": [
            {
              "dest": 8081,
              "alpn": "h2"
            },
            {
              "dest": 8080
            }
          ]
      },
      "streamSettings": {
          "network": "grpc",
          "security": "tls",
          "tlsSettings": {
            "certificates": [
              {
                "certificateFile": "/etc/ssl/v2ray/fullchain.pem",
                "keyFile": "/etc/ssl/v2ray/privkey.pem"
              }
            ]
          },
          "grpcSettings": {
              "serviceName": "example"
          }
      }
    },
 ```
 xray v1.4.0

**Comment by Dreamacro** [comment]
```yaml
  - name: gun
    server: 127.0.0.1
    port: 443
    type: trojan
    password: "example"
    network: grpc
    sni: example.com
    # skip-cert-verify: true
    udp: true
    gun-opts:
      grpc-service-name: "example"
```

it works

**Comment by fishg** [comment]
> ```yaml
>   - name: gun
>     server: 127.0.0.1
>     port: 443
>     type: trojan
>     password: "example"
>     network: grpc
>     sni: example.com
>     # skip-cert-verify: true
>     udp: true
>     grpc-opts:
>       grpc-service-name: "example"
> ```
> 
> ```yaml
>   - name: v2ray
>     server: 127.0.0.1
>     port: 443
>     type: vmess
>     uuid: b831381d-6324-4d53-ad4f-8cda48b30811
>     alterId: 32
>     cipher: auto
>     network: grpc
>     tls: true
>     servername: example.com
>     # skip-cert-verify: true
>     gun-opts:
>       grpc-service-name: "example"
> ```
> 
> it works

hi, "gun-opts" should be "grpc-opts" in vmess sample'.

**Comment by Dreamacro** [comment]
@fishg it's typo

File: component/gun/gun.go [hunk 1]
```diff
@@ -81,25 +94,34 @@ func (g *Conn) Read(b []byte) (n int, err error) {
...
+		return
+	}
 	n = copy(b, buf[1+protobufLengthLen:])
+	g.offset = n + 1 + int(protobufLengthLen)
+	g.buf = buf
 	return
 }
 
 func (g *Conn) Write(b []byte) (n int, err error) {
+	defer println("write")
```
[the remark this answers is not in the document]
Comment (Dreamacro) [reply component/gun/gun.go]: it's removed

**Comment by Dreamacro** [comment]
@ShinyGwyn If there are no other issues I would like to merge this PR for now?

**Comment by ShinyGwyn** [comment]
> If there are no other issues I would like to merge this PR for now?

I have no problem with that.

**Review by DuckSoft — APPROVED** [review]

**Comment by ShinyGwyn** [comment]
@Dreamacro There is a probability that it will appear under the current test
```
panic: runtime error: invalid memory address or nil pointer dereference
```

**Comment by Dreamacro** [comment]
fixed in 1355196b7c07a5d4f3319a2b0a357a8f7972a669

File: component/gun/gun.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+		cn := tls.Client(pconn, cfg)
+		if err := cn.Handshake(); err != nil {
+			pconn.Close()
+			return nil, err
+		}
+		state := cn.ConnectionState()
+		if p := state.NegotiatedProtocol; p != http2.NextProtoTLS {
+			cn.Close()
+			return nil, errors.New("http2: unexpected ALPN protocol " + p + "; want q" + http2.NextProtoTLS)
```
Comment (maskedeken) [inline component/gun/gun.go:168]: "want q"这个q是什么？

**Pull request merged**